### PR TITLE
String ractor

### PR DESCRIPTION
**Pull request by Saalvage**

### CODE REVIEW

**Comment by Saalvage** [comment]
Another idea was to make getHashCode private in order to force people into using the provided StringKey classes, but I could not realize that idea yet, because vertices seemingly try to implement their own map-like structure which relies on the hashcode. Why do they do that?

**Comment by juanjp600** [comment]
Vertices use the hashes to speed up comparisons and insertions, as they can have properties with arbitrary names. An unordered map would've been more efficient, but I was concerned about memory usage at the time. This will be changed to instead require that a shader be provided to a mesh, so it can extract the vertex layout from that shader and present vertices using that layout. This should address the memory concerns I was having before, to the extent that using the StringKey classes with a standard unordered map should still be lower cost.

File: Src/String/String.cpp [hunk 1]
```diff
@@ -165,30 +295,35 @@ const String PGE::operator+(const char* a, const String& b) {
...
+        }
+        _hashCodeEvaluted = true;
+    }
+    return _hashCode;
 }
 
 bool String::equals(const String& other) const {
-    if (other.byteLength()!=byteLength()) { return false; }
-    return other.getHashCode()==getHashCode();
+    if (_strByteLength >= 0 && other._strByteLength >= 0 && byteLength() != other.byteLength()) { return false; }
```
Comment (Saalvage) [inline Src/String/String.cpp:319]: I'm unsure what is really preferable here.

**Comment by Saalvage** [comment]
Makes sense, so after that's done do you think hiding getHashCode and by doing so forcing people to use the provided tools is something to consider?
I think this might actually serve to illustate why exposing it might lead to unfavourable implementations

File: Include/String/String.h [hunk 2]
```diff
@@ -53,51 +56,46 @@ class String {
...
         void wCharToUtf8Str(const wchar* wbuffer);
-        static wchar utf8ToWChar(const char* cbuffer);
         void reallocate(int size);
-        void reset();
         char* cstrNoConst();
     };
 
     bool operator==(const String& a, const String& b);
     bool operator!=(const String& a, const String& b);
     const String operator+(const String& a, const String& b);
```
Comment (Saalvage) [inline Include/String/String.h:97]: Why is a const String returned here?

Comment (juanjp600) [reply Include/String/String.h:97]: To prevent the following from being valid:
`str1+str2 = "thing";`

**Pull request merged**